### PR TITLE
Type check sentinel values

### LINKED ISSUE #1603: Future of MultiIndex
I think we can continue the discussion we have in #1426 about `MultiIndex` here.

In [comment](https://github.com/pydata/xarray/pull/1426#issuecomment-304778433) , @shoyer recommended to remove `MultiIndex` from public API.

I agree with this, as long as my codes work with this improvement.

I think if we could have a list of possible `MultiIndex` use cases here, 
it would be easier to deeply discuss and arrive at a consensus of the future API.

### PR DESCRIPTION
**Pull request by max-sixty**
- [ ] Closes #xxxx
 - [ ] Passes `black . && mypy . && flake8`
 - [ ] Fully documented, including `whats-new.rst` for all changes and `api.rst` for new API

I did this before I realized it only works on mypy master! But posting so no one else does it. We can wait to merge until the next version of mypy is out

### CODE REVIEW

File: xarray/core/utils.py [hunk 1]
```diff
@@ -701,3 +702,10 @@ def get_temp_dimname(dims: Container[Hashable], new_dim: Hashable) -> Hashable:
     while new_dim in dims:
         new_dim = "_" + str(new_dim)
     return new_dim
+
+
+class Default(Enum):
+    sentinel = 0
+
+
+__default = Default.sentinel
```
Comment (crusaderky) [inline xarray/core/utils.py]: I think this would be just as good?
```python
class Default:
    pass

__default = Default()
```

**Comment by crusaderky** [comment]
@max-sixty have a look at my changes and then feel free to merge

File: xarray/core/dataset.py [hunk 2]
```diff
@@ -890,11 +887,11 @@ def _replace(  # type: ignore
                 self._coord_names = coord_names
             if dims is not None:
                 self._dims = dims
-            if attrs is not self.__default:
+            if not isinstance(attrs, Default):
```
Comment (max-sixty) [inline xarray/core/dataset.py]: I feel like there's some Chesterton's fence here: is there some reason we had `self.` here?

**Comment by max-sixty** [comment]
Great, thanks @crusaderky 

I followed the Enum pattern because that was suggested in [python docs](https://github.com/python/typing/pull/240). But I agree your suggestion is a bit simpler. Maybe have a quick glance at those docs and unless you change your mind, hit the green button?

File: xarray/core/dataset.py (hunk 2 above)
Comment (crusaderky) [reply xarray/core/dataset.py]: I don't think so. ``self.__default`` is misleading - it was a class attribute to begin with (you could rewrite it as ``Dataset.__default``).

**Comment by crusaderky** [comment]
Ah now I understand what you meant with mypy master

**Pull request merged**